### PR TITLE
[DataGridPremium] Allow programmatically grouping non-groupable columns

### PR DESCRIPTION
**Pull request by MBilalShafi**
Handles `colDef.groupable` part of #10552

https://deploy-preview-11539--material-ui-x.netlify.app/x/react-data-grid/row-grouping/#read-only-row-groups

**ToDos:**
- [x] Update tests
- [x] Update docs
- [x] Update migration guide

### Breaking changes

- Non-groupable columns could now be grouped programmatically to create read-only row groups by controlling or initializing `rowGroupingModel`, or by updating the `rowGroupingModel` by API method `apiRef.current.setRowGroupingModel`.

### CODE REVIEW

File: packages/grid/x-data-grid-premium/src/components/GridColumnMenuRowGroupItem.tsx [hunk 1]
```diff
@@ -30,9 +30,10 @@ function GridColumnMenuRowGroupItem(props: GridColumnMenuItemProps) {
       onClick(event);
     };
 
-    const name = columnsLookup[field].headerName ?? field;
+    const groupedColumn = columnsLookup[field];
+    const name = groupedColumn.headerName ?? field;
     return (
-      <MenuItem onClick={ungroupColumn} key={field}>
+      <MenuItem onClick={ungroupColumn} key={field} disabled={!groupedColumn.groupable}>
```
Comment (MBilalShafi) [inline packages/grid/x-data-grid-premium/src/components/GridColumnMenuRowGroupItem.tsx:36]: I made an exception here as compared to the other `colDef` flags, i.e the item will be shown in the column menu for the grouping column in the disabled state for grouped columns that have `colDef.groupable = false`.

In this example company's `colDef` has `groupable = false`
image

The rationale is, that if a grouping is applied on a non-groupable column, there's no way for the users to check which column is grouped as the header says `Group`, so keep showing the disabled (read-only) column menu item will be a pointer that this specific column is grouped.

File: docs/data/data-grid/row-grouping/row-grouping.md [hunk 2]
```diff
@@ -191,6 +191,14 @@ In the example below, the `director` column can not be grouped. And in all examp
 
 {{"demo": "RowGroupingColDefCanBeGrouped.js", "bg": "inline", "defaultCodeOpen": false}}
 
+### Grouping non-groupable columns programmatically
+
+The columns with `colDef.groupable` set to `false` could be grouped programmatically by initialiing, passing in the `rowGroupingModel` prop, or updating `rowGroupingModel` by the API method `setRowGroupingModel` to generate read-only row groups. The users could not alter such groups from the UI.
```
Comment (cherniavskii) [inline docs/data/data-grid/row-grouping/row-grouping.md:196]: What do you think about structuring it similarly to https://github.com/mui/mui-x/pull/11574:

File: packages/grid/x-data-grid-premium/src/components/GridColumnMenuRowGroupItem.tsx (hunk 1 above)
Comment (cherniavskii) [reply packages/grid/x-data-grid-premium/src/components/GridColumnMenuRowGroupItem.tsx:36]: Makes sense to me

**Review by cherniavskii — APPROVED** [review]

File: docs/data/data-grid/row-grouping/row-grouping.md (hunk 2 above)
Comment (MBilalShafi) [reply docs/data/data-grid/row-grouping/row-grouping.md:196]: Nice idea 💡

**Pull request merged**